File: problems/206-reverse-linked-list/solution.py

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # Invariant: prev holds the already-reversed prefix, cur is the next node
        # to move onto it. Each step flips exactly one link.
        cur, prev = head, None
        while cur:
            # Tuple assignment evaluates the whole right side first, so cur.next
            # is read before it is overwritten — no temporary needed.
            cur.next, prev, cur = prev, cur, cur.next
        return prev  # cur ran off the end; prev is the original tail
# @lc code=end

File: problems/206-reverse-linked-list/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // Recurse to the tail first, then flip links on the way back up.
        // new_head is the original tail, passed through every frame unchanged.
        if (!head || !head->next) return head;
        ListNode* cur = head;
        ListNode* new_head = reverseList(cur->next);
        // cur->next is still the node ahead of us; point it back at cur,
        // then cut cur's forward link or the two nodes form a cycle.
        cur->next->next = cur;
        cur->next = NULL;
        return new_head;
    }
};
